base/reduce.jl

```diff
@@ -362,10 +367,105 @@ julia> reduce(*, [2; 3; 4]; init=-1)
 -24
 ```
 """
-reduce(op, itr; kw...) = mapreduce(identity, op, itr; kw...)
+function reduce(op, itr::T; kw...) where T
+    # Redispatch, adding traits
+    reduce(op, itr, eltype_or_default_eltype(itr), IteratorSize(T); kw...)
+end
+
+function reduce(op, itr, et, isize; kw...)
+    # Fallback: if nothing interesting is being done with the traits
+    # or the operation
+    return mapreduce(identity, op, itr; kw...)
+end
 
 reduce(op, a::Number) = a  # Do we want this?
 
+##### Operation specific reduce optimisations
+
+## vcat
+
+function reduce(::typeof(vcat), xs, T::Type{<:AbstractVector}, isize)
+    x_state = iterate(xs)
+    x_state === nothing && return T() # New empty instance
+    x1, state = x_state
+
+    ret = copy(x1)  # this is **Not** going to work for StaticArrays
```

Comment:
So did this use to work for StaticArrays and is breaking? What should be done to resolve this comment?

Reply:
Indeed it did, used to work, though it used to work a little weirdly.
Since the result type depended if the iterator was a `Array` or not.
```
julia> data = [(@SVector Int[1,2,3,4]), @SVector Int[1,2,3,4]]
2-element Array{SArray{Tuple{4},Int64,1,4},1}:
 [1, 2, 3, 4]
 [1, 2, 3, 4]

julia> reduce(vcat, data)
8-element Array{Int64,1}:
 1
 2
 3
 4
 1
 2
 3
 4

julia> reduce(vcat, (i for i in data))
8-element SArray{Tuple{8},Int64,1,8}:
 1
 2
 3
 4
 1
 2
 3
 4
```

We should do something to support it.
But I wasn't sure what, so I left the comment (It should have had a `#TODO`)

1. We can tighten this to apply only to `Vector` not to `AbstractVector`
2. We can check for `ismutable` and if not mutable, we can fallback to standard `reduce`
3. We can check for `ismutable` and if not mutable, we can fall back to using a `Array` for the return type

All 3 options leave it open to the package to define there own improved method for this, on there type.

Reply:
1 would be too bad. 3 would break the `reduce` interface. So 2 sounds like the best solution.